pg_search/src/writer/index.rs

```diff
@@ -18,19 +18,95 @@
 use super::{Handler, IndexError, SearchFs, WriterDirectory, WriterRequest};
 use crate::{
     index::SearchIndex,
+    postgres::transaction::IndexWriterLock,
     schema::{
         SearchDocument, SearchFieldConfig, SearchFieldName, SearchFieldType, SearchIndexSchema,
     },
 };
 use anyhow::{Context, Result};
-use std::collections::{
-    hash_map::Entry::{Occupied, Vacant},
-    HashMap, HashSet,
+use std::sync::Arc;
+use std::{
+    collections::{
+        hash_map::Entry::{Occupied, Vacant},
+        HashMap, HashSet,
+    },
+    path::Path,
+};
+use std::{fs, io, result};
+use tantivy::directory::{
+    DirectoryClone, DirectoryLock, FileHandle, FileSlice, Lock, WatchCallback, WatchHandle,
+    WritePtr,
+};
+use tantivy::{
+    directory::error::{DeleteError, LockError, OpenReadError, OpenWriteError},
+    IndexSettings,
 };
-use tantivy::{schema::Field, Index, IndexWriter};
+use tantivy::{directory::MmapDirectory, schema::Field, Directory, Index, IndexWriter};
 use tracing::warn;
 
+#[derive(Debug)]
+pub struct UnlockedDirectory(MmapDirectory);
+
+impl UnlockedDirectory {
+    pub fn open(directory_path: impl AsRef<Path>) -> Result<Self> {
+        if !directory_path.as_ref().exists() {
+            fs::create_dir_all(&directory_path).expect("must be able to create index directory")
+        }
+        Ok(Self(MmapDirectory::open(directory_path)?))
+    }
+}
+
+impl DirectoryClone for UnlockedDirectory {
+    fn box_clone(&self) -> Box<dyn Directory> {
+        self.0.box_clone()
+    }
+}
+
+impl Directory for UnlockedDirectory {
+    fn get_file_handle(&self, path: &Path) -> Result<Arc<dyn FileHandle>, OpenReadError> {
+        self.0.get_file_handle(path)
+    }
+
+    fn open_read(&self, path: &Path) -> result::Result<FileSlice, OpenReadError> {
+        self.0.open_read(path)
+    }
+
+    fn open_write(&self, path: &Path) -> result::Result<WritePtr, OpenWriteError> {
+        self.0.open_write(path)
+    }
+
+    fn atomic_write(&self, path: &Path, data: &[u8]) -> io::Result<()> {
+        self.0.atomic_write(path, data)
+    }
+
+    fn atomic_read(&self, path: &Path) -> result::Result<Vec<u8>, OpenReadError> {
+        self.0.atomic_read(path)
+    }
+
+    fn delete(&self, path: &Path) -> result::Result<(), DeleteError> {
+        self.0.delete(path)
+    }
+
+    fn exists(&self, path: &Path) -> Result<bool, OpenReadError> {
+        self.0.exists(path)
+    }
+
+    fn acquire_lock(&self, _lock: &Lock) -> result::Result<DirectoryLock, LockError> {
```

Comment:
I'd like to see some comments on this function that describe what we're doing, since it's wildly different from what Tantivy tells us to do.  Ya know, our pinky swear and such.